File: Two_Sum/Python/two_sum.py
from typing import List
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        """Return index of elements in the nums that can add up to targer.

        Args:
         nums: list of intigers
         targer: intiger

        Returns:
         list: list of two intiger.
        
        """

        # Approach two pointer
        # Time complexity O(n)
        # Space complexity O(n)

        # create a cache dict.

        cache_dict = {}

        for i in range(len(nums)):
            target_num = -1 * (- target + nums[i] )

            if nums[i] in cache_dict:
                return [cache_dict[nums[i]], i]
            cache_dict[target_num] = i
        return None


if __name__ == "__main__":
    r = Solution()
    result = r.twoSum([2,7,11,15], 9) 
    print(result)

File: Two_Sum/JavaScript/two_sum.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function(nums, target) {
    // Create a hash map to store the indices of numbers
    const numIndices = {};

    // Iterate through the array
    for (let i = 0; i < nums.length; i++) {
        // Calculate the complement needed to reach the target
        const complement = target - nums[i];

        // Check if the complement exists in the hash map
        if (complement in numIndices) {
            // Return the indices of the two numbers that add up to the target
            return [numIndices[complement], i];
        }

        // Store the current number's index in the hash map
        numIndices[nums[i]] = i;
    }

    // If no solution is found, return an empty array or handle it as needed
    return [];
};

// Define your main function
function main() {
    const nums = [2, 7, 11, 15];
    const target = 9;
    const result = twoSum(nums, target);
    console.log(result); // Output: [0, 1]
}

// Check if the script is being run directly (not imported as a module)
if (require.main === module) {
    main();
}
